src/executables/srpd_rotation.c

```diff
@@ -0,0 +1,409 @@
+/**
+ * @file srpd_rotation.c
+ * @author Ondrej Kusnirik <Ondrej.Kusnirik@cesnet.cz>
+ * @brief rotation utility for sysrepo-plugind
+ *
+ * @copyright
+ * Copyright (c) 2018 - 2022 Deutsche Telekom AG.
+ * Copyright (c) 2018 - 2022 CESNET, z.s.p.o.
+ *
+ * This source code is licensed under BSD 3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://opensource.org/licenses/BSD-3-Clause
+ */
+
+#define _GNU_SOURCE
+
+#include <assert.h>
+#include <dirent.h>
+#include <errno.h>
+#include <pthread.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <libyang/libyang.h>
+#include <sysrepo.h>
+
+#include "bin_common.h"
+#include "config.h"
+#include "srpd_common.h"
+#include "srpd_rotation.h"
+
+int
+srpd_rotation_init_cb(sr_session_ctx_t *session, void **private_data)
+{
+    int r = 0;
+    srpd_rotation_opts_t *opts;
+
+    opts = calloc(1, sizeof *opts);
+
+    /* create notification rotation change subscription */
+    if ((r = sr_module_change_subscribe(session, "sysrepo-plugind", "/sysrepo-plugind:sysrepo-plugind/notif-datastore/rotation/enabled",
+            srpd_notif_cb, opts, 0, SR_SUBSCR_ENABLED | SR_SUBSCR_DONE_ONLY, &opts->subscr))) {
+        SRPLG_LOG_ERR("srpd_rotation", "Failed to subscribe (%s)", r);
+        goto error;
+    }
+
+    /* create notification rotation state data change subscription */
+    if ((r = sr_oper_get_subscribe(session, "sysrepo-plugind", "/sysrepo-plugind:sysrepo-plugind/notif-datastore/rotation/rotated-files-count",
+            srpd_get_rot_count_cb, opts, 0, &opts->subscr))) {
+        SRPLG_LOG_ERR("srpd_rotation", "Failed to subscribe (%s)", r);
+        goto error;
+    }
+
+    *private_data = opts;
+    return 0;
+
+error:
+    sr_unsubscribe(opts->subscr);
+    free(opts);
+    return r;
+}
+
+void
+srpd_rotation_cleanup_cb(sr_session_ctx_t *session, void *private_data)
+{
+    srpd_rotation_opts_t *opts = (srpd_rotation_opts_t *)private_data;
+
+    sr_unsubscribe(opts->subscr);
+    opts->finish = 1;
+    srpd_notif_cb(session, 0, NULL, NULL, 0, 0, opts);
```

Comment:
Why delegate thread join to the change callback and not perform it in this function?

Reply:
True, I changed it